### ISSUE TITLE
Ships should not boot in two places

### ISSUE THREAD

**Issue by galenwp** [urbit/urbit#5867]
(I fully admit this isn't a 'bug' per se, but it's something I think we should probably fix. I'm open to moving this discussion elsewhere if it's better off there.)

At present, when you try to start a ship with the same key revision in a new place it will (a) brick the original ship and (b) never successfully boot in the new place. The only way to resolve this is, of course, to breach—which is unfortunate for the user (who probably didn't understand this constraint to begin with).

I see two possible solutions:

A - 
At the very least, we could (on first boot of a non-comet), require that the user type something to confirm that their ship is not running anywhere else. This is probably the easiest path that will catch a high percentage of cases, but it's not perfect.

B - 
Another route could be to somehow try to ping the node and see if it's already online. I can see a bunch of problems with this approach, and I'm not sure what the obvious architectural solution is. Does that ping go through the parent? How do you authenticate those requests? Since you're booting the ship with a valid key, we could have you sign a single message to check whether the same identity is online.

Or perhaps there's something I haven't thought of. The main desired output is just that people can't try to boot, kill their other ship and be forced to breach with no user feedback.

**Comment by jeremytunnell** [comment]
Another alternate idea:

This lock file could be local, and generated from a fingerprint of the local machine.  We could hash a few pieces of the environment or something.  If booting is attempted where the hash doesn't match we display a warning message.

**Comment by belisarius222** [comment]
On dual-booting your urbit.

There are two kinds of problems that are basically the same problem technically but occur in two different ways. And there's been some interest recently in addressing this very long-standing issue with Urbit. It's been around since the very beginning.

So the issue is that you can't run two copies of your ship simultaneously.

There are two ways this can happen: one is that you boot up your ship, you run it for a little while, but then you forget that you've booted it. And so you then boot it up a second time, maybe on the same machine in a different folder, or on a different machine.

And so then, when that happens, the second copy won't be able to communicate with the sponsor, because the sponsor remembers the sequence numbers from the first instance of the ship that you booted.

And anyone else that you've talked to also, you won't be able to communicate with them, or or the communication will be haphazard, at best.

The same kind of situation can happen if you attempt to move your ship, but don't turn off the old copy. So let's say I'm running my ship on my laptop, and then I want to move it up to a cloud server. So I shut it down, zip it up, copy the zipped folder to the cloud.

But then I restart my ship on my laptop and also start the ship on the cloud, and so now I have two copies again.  This is a little bit different because it's not a fresh boot.

The copy I restart first will work, mostly. But if one copy talks to some people and the other one talks to some other people, then then whoever gets ahead first wins. It's vaguely analogous to a longest chain algorithm.

It seems like people are actually hitting both of these. I think that the more ships people have, the more likely it is that they'll do this.  Also the more that they move it around, the more likely it is, and it's just something that people aren't used to.  Most programs that people have run don't have this property.  Urbit is probably the only program they've ever encountered that has this property, so they're not expecting it and they're not on the lookout for it.

And it's easy to mess up if you're moving a lot of ships around, starting and stopping a lot of different ships.  So alright, so how can we address this?

Galen suggested adding a warning when you boot up your ship like hey, are you sure you haven't booted this somewhere else?  I think that's a good idea.

That's definitely a good first step.  There’s no reason not to do that in my opinion.  It would also be nice to have a command line flag to say don't ask me, just start it.

But then the second option is basically to try to figure out has this ship been booted before? And that's not exactly the question that needs to be answered.

A more precise framing is something like: if I boot this ship now, which is either fresh or not, do the sequence numbers that this ship has in its pier folder (which will be zero if it's fresh), do those sequence numbers match with the sequence numbers that peers have?  Peers, meaning ships that this ship has talked to.

You could imagine trying to ask everybody that you've ever talked to, but you shouldn’t need to because every ship pings its sponsor every 30 seconds.

So that's going to be approximately as up to date as anything else. It’s very unlikely that your sponsor doesn't have the latest understanding of how many times you've talked to it.

I think you should be able to get away with just asking the sponsor, and it's actually better to ask the sponsor than to ask the ship directly because the ship might have been shut down.

And the thing you want to check is really not whether the ship is currently running, but whether the ship whether the sequence numbers other people have for this ship match what we have in our pier.

So we need to ask the sponsor somehow: how many times have you been pinged by this sponsee?  So then we need to figure out how best to ask this question over the network before starting a ship.

It’s not particularly clear to me, but I can think of a few different options.  The three basic options are ames, remote scry, and HTTP.

Remote scry isn't quite ready yet, and even even once we've launched basic remote scry, trying to ask for the latest information about something isn't something that's built into the scry namespace.

I have some ideas about how to build that kind of query. But we haven't built that yet.  Once we have that kind of query, you could ask the sponsor: what's the latest that you've heard from this sponsee?  But that's not on the table at the moment.

So that leaves Ames and HTTP.

With Ames, the problem is that if you try to send a request, as our current ship, that's a stateful request.

The other ship will need to update its state, so if the other copy of your ship is trying to talk to the sponsor, then they could clobber each other. Or if you do this multiple times, then because Ames sequence numbers are all permanent, it might fail if you try it again.

I can think of two ways of trying to deal with this. The easy way is to say, don't use your ship, spawn up a comet and have the comet ask the sponsor, what's the sequence number?

It'll have to do like a little pre boot thing into the pier. If there's an existing Arvo state, then it'll have to get it running, but not give it any events, just scry into it and ask it what its latest sequence number is for the sponsor (on the ping bone, which is bone 1).

If there's no pier, it's a fresh ship, then we're comparing the sponsor’s sequence number against zero.


It’s easy to expose a sponsee’s latest ping sequence number on the sponsor, and it's not security sensitive information, so it's fine to just expose it publicly.

The other, less easy way of asking the sponsor using Ames (and this might or might not work, I'm not sure) is to use some sort of temporary or random bone.

A bone is a number representing a connection between ships, called a “flow”.  A pair of ships can have multiple flows going simultaneously. They're like sockets.

A bone is very much like Ames’s equivalent of a port, except that they just increment.  Each time you open a new flow, it just increments the bone.  You’re not listening at a particular bone like you would in Unix.

So what you'd want to do is either pick a random bone that's a very large number, so it’s unlikely to collide with other bones, or there are two other options.

One is to create a new bone, and then delete it once this whole thing goes through, by closing the flow.  We can do this in Ames as of the gall request queue fix.

Once the flow has been closed, you can imagine reopening it, because closing it deletes the sequence numbers. So you could do that, although there’s a risk that the connection terminates before the flow closes, so the sponsor ends up storing a bunk flow that can’t be cleaned up easily.

And then there's a third option, which is basically that you have some sort of subset of the bone space reserved just for this particular piece of functionality.  Pick a bone within this part of the space, and you use that.

Maybe you also delete it, or maybe it deletes automatically on a timer or something.  You could imagine having different semantics for this interaction just to keep it easy.


Or maybe it's not even by bone and it's done some other way.  You could even have a special Ames packet type that doesn't have a bone that just says, hey, please tell me what the latest sponsee ping number is.

Because it's a single packet request, and a single packet response. It's not too hard to imagine having a separate protocol for this. I send you this packet and it has a request saying, what's the latest sequence number for the sponsee, and the sponsor just sends back a response.  That response is a single packet that contains a sequence number.  You can get different numbers, potentially, if you send this request multiple times, but that's okay.

This way the sponsor doesn't have to remember that it got the request, because it doesn't have to update any of its state in order to respond.  It could remember, but it doesn't have to, which is convenient.

This is really like a bespoke network protocol.  It’s probably inside of the Ames protocol, but maybe not. There's a whole design space here of ways that you could try to imagine sending packets back and forth over UDP using real urbit identities, but in a way that doesn't break anything — you don't clobber any state or have some sort of competition between duplicate instances of a ship.

That concludes the second set. of options, which is do this Ames. Again, there might be something viable in there, I'm not sure.

The third option is to do this over HTTP, but it’s not a good option.

You’d send an HTTP request to the sponsoring ship, and it would send back a response.  The nice thing about that is it’s easy to do from from the runtime, which already makes HTTP requests for various things.

And because it's HTTP, there aren't permanent sequence numbers like in Ames, so you don't run into this issue with clobbering state by asking this question.

But the problem with doing it over HTTP is that not all ships expose HTTP servers, and not necessarily at a deterministic address.  Not even galaxies do.  They have to be available over UDP at a specific port at some domain that they specify through Ethereum. 

But your star doesn't have to be available over HTTP anywhere.  Even if we stipulated that a star must be available over HTTP, your sponsor might actually be a planet.  You might be moving a moon and doing doing this dual-boot check on your moon.

In that case, your sponsor is a planet, and there will never be a requirement that a planet must be available at some HTTP endpoint.

So HTTP actually doesn't really work.

I think of these options, the best one is the comet.

The problem is that it takes a while to boot a comet, but we could spawn a thread pool to mine for comets, and then it would be much less of a problem.

Also, if you move your ship from one place to another and it takes a couple of minutes to boot it back up, that’s not a big deal.

There probably is some answer in the bespoke protocol part of the design space.  There probably is some bespoke protocol that we could build that would work well for this that doesn't require spawning a comment. We would have to come up with that idea and spec it out, and then build it.

It could be similar to the the comet self-attestation protocol, where it sends its own key and signs it.

Maybe we could do something similar here, where you send your key and sign it, and the sponsor ship will respond with the sequence number that it has for you, or something like that.

I think the key property to take advantage of, if we go the route of a bespoke protocol, is the fact that we can do this whole check in one packet roundtrip, which should mean that we don't need to track sequence numbers at all if we did it right, and that could potentially simplify things quite a bit.

Transcribed by https://otter.ai

**Comment by jeremytunnell** [comment]
Why don't we look at this from a user perspective instead of a technical perspective?

What if we added a feature that was something like "move this pier".  

When somebody activated this feature, it would essentially flag the current pier as no longer valid and it would copy the pier into a zip file or some type of archive that could be moved elsewhere.  

Then there would be an "activate this pier" feature, that would simply put the contents of the archive into a new location.  (./urbit activate-pier archive.xyz)

This would completely solve the "run twice after copy" problem because every old copy would be flagged as invalid.  

If you wanted to make it extra solid, you could tie the pier to the local environment in some way that it simply would not work if you just copied it..."ERROR:  this pier is tied to x computer, if you want to move it use the 'move pier' feature"

This is extremely easy, could be hammered out very quickly and foolproof for half of the problem.

**Comment by Fang-** [comment]
fwiw, someone from support indicated that the most common case of "booted it twice" issues they see is users thinking they can just turns their urbit "off and on again" by starting from scratch. The suggestions around showing a message/prompt during boot could help here, as a form of "inline education".

**Comment by jeremytunnell** [comment]
I get you, but if we're going to fix this, let's fix it.  

When I ran UX at Stack Overflow, we had a general rule that adding a warning message was a last resort.  It is better to make it so the user can't screw up in the first place.

**Comment by ohAitch** [comment]
And I suppose they can fix the converse case by rotating their keys, fair
enough

**Comment by forfel-norfel** [comment]
Sorry in advance if I've misread this, but it seems like y'all are trying to come up with a solution to keep people from "double spending" their piers and pier backups. Some guy came up with a neat way to do that back in 2008. It's kinda obscure. You probably haven't heard of it.

Seriously though, this doesn't require a blockchain or anything, but it's the exact same problem. It absolutely cannot be done offline. It requires either a centralized oracle or some sort of coordinated effort to determine the truth. The user here isn't intentionally behaving like a malicious actor, but they're still effectively a malicious actor.

A prompt to warn the user is a simple solution, but like every other warning out there, it'll be worthless if the times that it appears don't almost perfectly overlap with the times that something bad is about to happen. If the warning appears every time someone boots their existing planet, it will just waste time and prevent nothing.

How can you determine that it should or shouldn't be displayed then? You can't. What's the difference between a pier and a backup of that pier? Either you allow people to make proper backups that work exactly as well as the original, or you're just recreating the software hell we're all trying to escape. If booting an urbit is affected by environment details that aren't well documented and easily saved and reproduced, you're just undermining the whole project.

It also can't be done reliably anyway. You can easily catch a few common scenarios, like maybe the path to the pier directory changes, but that'll lead to a lot of false positives and not do anything about disk images or VM clones or people who just aren't very creative about folder names. Anything more reliable will require a more advanced solution, which is basically just old video game disc copy protection schemes or rootkits or other DRM solutions.

Sorry, but it just isn't possible to give people freedom while also offering a feature that requires you to have some control over every computer they touch. There's simply no practical offline solution here because this is literally the double spending problem.

So how do you solve this online? It's a relatively straightforward problem at that point in terms of coordination. Record timestamps for last startup and shutdown, or whatever other tiny bits of metadata allow the logic to be most efficient and reliable, and because you're only trying to defeat an unintentional double spend and nobody else has anything to gain, trust isn't a concern. (I'm sure some absurd security exploit will be discovered years from now that somehow takes advantage of this. Please don't judge me from the future. 2022 was a much simpler time.)

The challenge then becomes making it perform well and be unlikely to fail. If a ship can't quickly determine whether it should be running, then we're back to prompting the user for no reason. Maybe I haven't fully thought this through, but it seems like using DNS lookups to handle this task would be an elegant, reliable, and performant approach. DNS is already an essential external dependency and it's lightweight and fast. Because it's out of band, adding this feature shouldn't create any additional complexity for communication between ships. There are tons of ways to abuse DNS, but I'm imagining that this could be handled as part of a lookup that already needs to be made. Rather than simply requesting an A record for a star or galaxy, the ship could perform a lookup for a subdomain that encodes whatever information needs to be shared. Since these requests would always be unique, they would bypass any caching and always reach the authoritative name server. The name server could then log the request and respond with either the star/galaxy A record or something different that informs the ship of a potential problem.

Ultimately, this just seems like a temporary workaround though. It seems very important to fix the underlying problem so that booting multiple copies of a ship is only a temporary problem that can easily be fixed, a bit of data loss doesn't require a breach, and booting from an old backup is not only harmless but even a viable way to get a ship running. I'm confident that most conflicts and data losses could be overcome with the right rebuild processes and a bit of manual guidance, so that an urbit can almost always be returned to a consistent and functional state. I'm sure actually developing that rebuild process will be challenging, but I don't see any reason why it wouldn't be possible, and it seems like the sort of thing that's going to need to happen eventually.

**Comment by belisarius222** [comment]
You’re right that double-booting could be considered a Byzantine fault, so solving it requires network-wide coordination. I also share your skepticism that user-facing warnings will make a big enough dent in the problem, which you articulated well, and I agree with you about security.

Your approach to a solution ignores something important, though: the totally ordered sequence of pings to the sponsor that a ship performs every 30 seconds, and which the sponsor records persistently.

The star can be thought of as a “trusted third party” (or the sole “authority” in a “proof of authority” consensus, if you prefer) for resolving disputes between two potential instances of a ship. As I described in my earlier post, since we can check this information before starting to run a ship, the runtime could compare the local pier’s latest ping sequence number with the one stored on the star. If another instance of the ship is running, the local pier will have a lower number than the star, so we know we shouldn’t try to run this instance.

(Due to possibility of packet loss for acks, the local pier could have a last-acked number that’s one less than the star; this is one of a few imperfections with this design, but the likelihood that you started two instances within 30 seconds of each other is quite low, so we should treat off-by-one as ok to start this pier.)

I have yet to see a direct response to my proposal. If it’s unclear, I’d be happy to try to explain it better.

—
~rovnys-ricfer
https://urbit.org

**Comment by jeremytunnell** [comment]
~rovnys-ricfer You provided several options, but argued against all of them but one.  Your "comet" option just seems quite complex and also would not fully solve the problem (What if the star isn't online when you're trying to boot?)

forfel-norfel I am unclear about your proposal.  Could you go into more detail about how it would actually work?  Is there some urbit DNS system, or are you suggesting a user would need to add a domain record at their DNS provider before they could boot?

As far as I can see, the solution I've suggested would fix 90% of the problem and would probably only take a dev less than a day to implement.  

I would love to fix 100% of the problem, and if somebody has a suggestion on how to do that, let's get there.  But in the absence of a complete solution, it seems to me wise to just spend one day on it and solve 90% of it and move on.

**Comment by belisarius222** [comment]
I can think of three ways a user ends up double-booting, ordered by descending prevalence:
1) User does not understand that there should only be one copy of their ship running
2) User forgets that the ship has already been booted somewhere else, and tries to boot it again
3) User, while trying to move the ship, forgets whether the old ship is also running

Scenario 1) is most aptly solved with a warning, as @Fang- suggested, since it's basically just an educational problem.  A warning doesn't help much with 2) or 3), though.

Adding import and export subcommands to the urbit executable addresses 1) and simple cases of 3), but not 2).

Asking the sponsor before boot solves all three unless your star is offline or the two piers start within a minute or so of each other -- I think those are the only cases where that wouldn't work.  Your star will almost always be online, that's its job.  A bad script in a hosting provider could boot two copies of a ship at the same time, but it's pretty unlikely a user would do that.  So asking the sponsor doesn't address every single case, but I think it's about as close as we can get.

I'm pretty sure that as @Fang- mentioned, the most pressing part of this problem by far is users just not understanding that it would be a problem to run two copies of the ship.  This suggests we should add a warning and make the user hit enter to confirm boot, even though as @forfel-norfel wrote, it doesn't prevent accidents.  If we do add a warning, we'll also want a command-line flag like `--yes` to suppress the interactivity when booting with a script.  I think this is the first step we should take toward addressing the double-boot issue.

In my opinion, we should add special urbit subcommands for importing and exporting piers.  This would help portability for large piers, and it will help prevent double-booting in scenarios 1) and 3), if we mark a pier as you suggested, @jeremytunnell.  I think this would be more work than the warning, but I'm not sure.

I think it's likely we'll want to build something that checks with the sponsor before boot, but not now.  @jeremytunnell suggested using lockfiles, @forfel-norfel proposed DNS lookups, and I proposed a few Urbit-based solutions.  It's almost always less work to build something inside Urbit than outside it.  It's also easier for us to deploy something using Ames and remote scry than something using DNS and HTTPS -- does the user have a public IP? TLS cert? do you need to follow redirects? support HTTP/2, etc.

Booting up an ephemeral comet just to check the sponsor for a previous boot is indeed slow and heavyweight, though.  Once we have the remote scry protocol deployed, especially with a good solution for [`%pine` requests](https://gist.github.com/belisarius222/95ad2a5f650ce90a61bb575f9ce1bcba), checking with the sponsor should get substantially easier, so I imagine we'll reevaluate then.

**Comment by jeremytunnell** [comment]
@belisarius222 I just have a major concern about your comet idea in that i'm not sure I'll have the resources on any of my stars to run a comet.  I'm currently on 1gb VPSs with swap.

What I'm sensing is that what should happen eventually is we either fix the sequence problem (big job), or we fix the double booting problem for good using your scry idea or perhaps the DNS idea (with some fleshing out).  

However, it's also my sense that none of these options are really ready to be done now, and I'm sure devs have many other priorities.  

So my suggestion is let's put a very small amount of time into some interface changes and small features to reduce this problem by 80%.  I believe what I've outlined will roughly do that.  

If we can come to some sort of agreement on that, I'm happy to write a spec and put some real thought into covering all of the bases.
